functions/snippets/src/main/java/PubSub.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START functions_tips_infinite_retries]
+import java.sql.Timestamp;
+// [START functions_tips_retry]
+import java.text.ParseException;
+import java.util.Map;
+
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+public class PubSub {
+  // Use GSON (https://github.com/google/gson) to parse JSON content.
+  private Gson gsonParser = new Gson();
+
+  // [END functions_tips_infinite_retries]
+  // [START functions_tips_retry]
+  public String retryPubSub(PubSubMessage message) throws Exception, ParseException {
+    JsonObject body = gsonParser.fromJson(message.data, JsonObject.class);
+    
+    boolean retry = false;
+    if (body.has("retry")) {
+      retry = body.get("retry").getAsBoolean();
+    }
+    
+    if (retry) {
+      throw new Exception("Retrying...");
```

Comment:
This doesn't seem right. I think what we should have is that if you want to trigger a retry then you throw an exception, which will result in an HTTP 500 response, and that should trigger the retry logic. If you don't want to retry then you just return. But here we're throwing an exception in both cases.

Reply:
We should probably throw something more specific than `Exception` as well.

Reply:
- I'm following the canonical sample for this snippet:
  - https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/master/functions/tips/index.js#L166
- I've changed the `Exception` to a `RuntimeException`.